scylla-cdc/src/stream_reader.rs

```diff
@@ -168,21 +172,71 @@ impl StreamReader {
         values: &[CqlValue],
         consumer: &mut Box<dyn Consumer>,
     ) -> anyhow::Result<()> {
-        let mut rows_stream = self
-            .session
-            .execute_iter(query_base.clone(), &values)
-            .await?;
-
-        let schema = CDCRowSchema::new(rows_stream.get_column_specs());
+        let mut sleep_after_timeout = BASIC_TIMEOUT_SLEEP_MS;
 
-        while let Some(row) = rows_stream.next().await {
-            consumer
-                .consume_cdc(CDCRow::from_row(row?, &schema))
-                .await?;
+        let mut next_state = None;
+        let mut page_no = 0;
+        loop {
+            let state_clone = next_state.clone();
+            let query_res = self
+                .session
+                .execute_paged(query_base, values, next_state)
+                .await;
+            match query_res {
+                Ok(x) => {
+                    sleep_after_timeout = BASIC_TIMEOUT_SLEEP_MS;
+                    page_no += 1;
+                    if let Some(rows) = x.rows {
+                        let schema = CDCRowSchema::new(&x.col_specs);
+
+                        for row in rows {
+                            consumer.consume_cdc(CDCRow::from_row(row, &schema)).await?;
+                        }
+                    }
+                    match x.paging_state {
+                        Some(state) => next_state = Some(state),
+                        None => break,
+                    }
+                }
+                Err(
+                    QueryError::TimeoutError | QueryError::DbError(DbError::ReadTimeout { .. }, _),
+                ) => {
+                    Self::print_timeout_warning(values, sleep_after_timeout, page_no).await;
+                    sleep(time::Duration::from_millis(sleep_after_timeout as u64)).await;
+                    sleep_after_timeout *= TIMEOUT_FACTOR;
+                    if sleep_after_timeout >= self.config.sleep_interval.as_millis() {
+                        sleep_after_timeout = self.config.sleep_interval.as_millis();
+                    }
+
+                    next_state = state_clone;
+                }
+                Err(e) => {
+                    return Err(anyhow::Error::new(e)
+                        .context("Session returned an error while fetching CDC rows."));
+                }
+            }
         }
-
         Ok(())
     }
+
+    async fn print_timeout_warning(values: &[CqlValue], backoff: u128, page_no: u64) {
+        if enabled!(tracing::Level::WARN) {
+            let ids_str = values[0..values.len() - 2]
+                .iter()
+                .map(|x| format!("0x{}", hex::encode(x.as_blob().unwrap())))
+                .join(", ");
+            let begin = values[values.len() - 2].as_duration().unwrap();
+            let end = values[values.len() - 1].as_duration().unwrap();
+
+            warn!(
+                "WARN: Timeout while fetching CDC rows for streams: {}.\n\
```

Comment:
`WARN:` at the beginning of the string is redundant. The backend responsible for printing/storing messages logged with the `tracing` crate will usually include information about the verbosity.

Reply:
Quite funny - I was inspired by [this code](https://github.com/piodul/scylla-cdc-rust/blob/ae8ccb5f8d03b1177b55c89c2b91dbf4176cb827/scylla-cdc/src/checkpoints.rs#L42) from the repository.

Reply:
Well, it slipped past my review. The code you pointed to should be fixed, too (and other instances of such problem, if there are any).

Reply:
I fixed this in #112

Reply:
Even though you fixed the other place, here you still have a redundant `WARN` in the string :(